packages/marketplace/contracts/transfer-manager/TransferManager.sol

```diff
@@ -119,120 +119,80 @@ abstract contract TransferManager is ERC165Upgradeable, ITransferManager {
-    /// @param from owner of the Asset to transfer
-    /// @return How much left after transferring royalties
+    /// @return How much left after paying fees
     function _transferRoyalties(
-        LibAsset.AssetType memory paymentAssetType,
-        LibAsset.AssetType memory nftAssetType,
-        LibPart.Part[] memory payouts,
-        uint256 rest,
-        uint256 amount,
-        address from
+        DealSide memory paymentSide,
+        DealSide memory nftSide,
+        uint256 rest
     ) internal returns (uint256) {
-        LibPart.Part[] memory royalties = _getRoyaltiesByAssetType(nftAssetType);
+        LibPart.Part[] memory royalties = _getRoyaltiesByAssetType(nftSide.asset.assetType);
+        uint256 len = royalties.length;
+        address creator = _getCreator(nftSide.asset.assetType);
 
-        address creator = _getCreator(nftAssetType);
-        if (creator != address(0) && payouts[0].account == creator) {
+        // When the creator buys his own tokens he doesn't pay royalties to anybody.
+        if (creator != address(0) && nftSide.to == creator) {
+            require(royalties[0].value <= 5000, "Royalties are too high (>50%)");
             require(royalties[0].value <= 5000, "Royalties are too high (>50%)");
             return rest;
         }
-        if (royalties.length == 1 && royalties[0].account == payouts[0].account) {
+
+        // This is an optimization to avoid doing two transfers
+        // in the case that the buyer has royalties over the token he is buying in this tx.
+        if (len == 1 && royalties[0].account == nftSide.to) {
             require(royalties[0].value <= 5000, "Royalties are too high (>50%)");
             return rest;
         }
 
-        (uint256 result, uint256 totalRoyalties) = _transferFees(paymentAssetType, rest, amount, royalties, from);
-        require(totalRoyalties <= 5000, "Royalties are too high (>50%)");
-        return result;
-    }
+        uint256 totalRoyalties;
+        LibAsset.Asset memory payment = LibAsset.Asset(paymentSide.asset.assetType, 0);
+        for (uint256 i; i < len; i++) {
+            LibPart.Part memory r = royalties[i];
+            totalRoyalties = totalRoyalties + r.value;
+            (rest, payment.value) = _subFeeInBp(rest, paymentSide.asset.value, r.value);
+            if (payment.value > 0) {
+                transfer(payment, paymentSide.from, r.account);
+            }
+        }
 
-    /// @notice calculates royalties by asset type.
-    /// @param nftAssetType NFT Asset Type to calculate royalties for
-    /// @return calculated royalties (Array of LibPart.Part)
-    function _getRoyaltiesByAssetType(LibAsset.AssetType memory nftAssetType) internal returns (LibPart.Part[] memory) {
-        (address token, uint256 tokenId) = abi.decode(nftAssetType.data, (address, uint));
-        return royaltiesRegistry.getRoyalties(token, tokenId);
+        require(totalRoyalties <= 5000, "Royalties are too high (>50%)");
```

Comment:
We have this require repeated a lot of this _transferRoyalties` function. Is there a way to code this to only check it once?

Reply:
I don´t like the alternatives, we can merge the two if blocks but we need to be sure about the conditions first.